File: frontend/src/services/knowledgeBase.ts
import api from './api';

export interface KnowledgeBase {
  id: number;
  name: string;
  description?: string;
  vector_store_type: 'elasticsearch' | 'pgvector';
  vector_store_config_id?: number | null;
  embedding_provider: 'openai' | 'claude' | 'ollama';
  embedding_model: string;
  chunk_size: number;
  chunk_overlap: number;
  is_public: boolean;
  user_id: number;
  document_count: number;
  total_chunks: number;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeBaseCreate {
  name: string;
  description?: string;
  vector_store_type?: 'elasticsearch' | 'pgvector';
  vector_store_config_id?: number | null;
  embedding_provider?: 'openai' | 'claude' | 'ollama';
  embedding_model?: string;
  chunk_size?: number;
  chunk_overlap?: number;
  is_public?: boolean;
}

export interface EmbeddingProvider {
  id: string;
  name: string;
  models: EmbeddingModel[];
}

export interface EmbeddingModel {
  id: string;
  name: string;
  description: string;
  dimensions: number;
}

export interface VectorStoreConfig {
  id: number;
  name: string;
  type: string;
  url: string;
  status: string;
}

export interface Document {
  id: number;
  knowledge_base_id: number;
  filename: string;
  file_type: string;
  file_size: number;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  error_message?: string;
  chunk_count: number;
  character_count: number;
  uploaded_at: string;
  processed_at?: string;
}

export interface Chunk {
  id: number;
  document_id?: number;
  chunk_index: number;
  content: string;
  character_count: number;
  metadata?: any;
  created_at: string;
}

export interface ChunkListResponse {
  total: number;
  chunks: Chunk[];
}

export interface SearchResult {
  content: string;
  score: number;
  document_id: number;
  document_name: string;
  chunk_index: number;
  metadata?: any;
}

// 获取知识库列表
export const getKnowledgeBases = async (): Promise<KnowledgeBase[]> => {
  const response = await api.get('/knowledge-bases');
  return response.data;
};

// 创建知识库
export const createKnowledgeBase = async (data: KnowledgeBaseCreate): Promise<KnowledgeBase> => {
  const response = await api.post('/knowledge-bases', data);
  return response.data;
};

// 获取知识库详情
export const getKnowledgeBase = async (id: number): Promise<KnowledgeBase> => {
  const response = await api.get(`/knowledge-bases/${id}`);
  return response.data;
};

// 更新知识库
export const updateKnowledgeBase = async (
  id: number,
  data: Partial<KnowledgeBaseCreate>
): Promise<KnowledgeBase> => {
  const response = await api.put(`/knowledge-bases/${id}`, data);
  return response.data;
};

// 删除知识库
export const deleteKnowledgeBase = async (id: number): Promise<void> => {
  await api.delete(`/knowledge-bases/${id}`);
};

// 上传文档
export const uploadDocument = async (kbId: number, file: File): Promise<Document> => {
  const formData = new FormData();
  formData.append('file', file);
  
  // 注意：不要手动设置 Content-Type，让浏览器自动添加 boundary 参数
  const response = await api.post(`/knowledge-bases/${kbId}/documents`, formData);
  return response.data;
};

// 获取文档列表
export const getDocuments = async (kbId: number): Promise<Document[]> => {
  const response = await api.get(`/knowledge-bases/${kbId}/documents`);
  return response.data;
};

// 删除文档
export const deleteDocument = async (docId: number): Promise<void> => {
  await api.delete(`/knowledge-bases/documents/${docId}`);
};

// 获取文档分块列表
export const getDocumentChunks = async (docId: number): Promise<ChunkListResponse> => {
  const response = await api.get(`/knowledge-bases/documents/${docId}/chunks`);
  return response.data;
};

// 批量导入文档
export const batchImportDocuments = async (
  kbId: number,
  directoryPath: string,
  fileExtensions?: string[]
): Promise<any> => {
  const response = await api.post(`/knowledge-bases/${kbId}/batch-import`, {
    directory_path: directoryPath,
    file_extensions: fileExtensions,
  });
  return response.data;
};

// 搜索知识库
export const searchKnowledgeBases = async (
  query: string,
  topK?: number,
  knowledgeBaseIds?: number[]
): Promise<{ query: string; results: SearchResult[]; total: number }> => {
  const response = await api.post('/knowledge-bases/search', {
    query,
    top_k: topK,
    knowledge_base_ids: knowledgeBaseIds,
  });
  return response.data;
};

// 机器人关联知识库
export const associateKnowledgeBases = async (
  robotId: number,
  knowledgeBaseIds: number[]
): Promise<any> => {
  const response = await api.post(`/robots/${robotId}/knowledge-bases`, {
    knowledge_base_ids: knowledgeBaseIds,
  });
  return response.data;
};

// 获取机器人的知识库
export const getRobotKnowledgeBases = async (robotId: number): Promise<any> => {
  const response = await api.get(`/robots/${robotId}/knowledge-bases`);
  return response.data;
};

// 取消机器人与知识库的关联
export const disassociateKnowledgeBase = async (
  robotId: number,
  kbId: number
): Promise<void> => {
  await api.delete(`/robots/${robotId}/knowledge-bases/${kbId}`);
};

// 获取嵌入模型提供商
export const getEmbeddingProviders = async (provider?: string): Promise<{ providers: EmbeddingProvider[] }> => {
  const url = provider ? `/providers/embeddings?provider=${provider}` : '/providers/embeddings';
  const response = await api.get(url);
  return response.data;
};

// 获取向量存储配置
export const getVectorStoreConfigs = async (): Promise<{ vector_stores: VectorStoreConfig[] }> => {
  const response = await api.get('/providers/vector-stores');
  return response.data;
};

File: backend/app/schemas/knowledge_base.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime

from app.models.knowledge_base import VectorStoreType, DocumentStatus
from app.ai.models import EmbeddingModel, EmbeddingProvider


# ============= 知识库 Schemas =============

class KnowledgeBaseBase(BaseModel):
    """知识库基础模型"""
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    vector_store_type: VectorStoreType = VectorStoreType.PGVECTOR
    vector_store_config_id: Optional[int] = None  # 外部向量存储配置ID（NULL表示使用系统库）
    embedding_provider: EmbeddingProvider = EmbeddingProvider.OPENAI
    embedding_model: EmbeddingModel = EmbeddingModel.OPENAI_SMALL
    chunk_size: int = Field(default=500, ge=100, le=2000)
    chunk_overlap: int = Field(default=50, ge=0, le=500)
    is_public: bool = False


class KnowledgeBaseCreate(KnowledgeBaseBase):
    """创建知识库"""
    pass


class KnowledgeBaseUpdate(BaseModel):
    """更新知识库"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    vector_store_type: Optional[VectorStoreType] = None
    vector_store_config_id: Optional[int] = None
    embedding_provider: Optional[EmbeddingProvider] = None
    embedding_model: Optional[EmbeddingModel] = None
    chunk_size: Optional[int] = Field(None, ge=100, le=2000)
    chunk_overlap: Optional[int] = Field(None, ge=0, le=500)
    is_public: Optional[bool] = None


class KnowledgeBaseResponse(KnowledgeBaseBase):
    """知识库响应模型"""
    id: int
    user_id: int
    es_index_name: Optional[str]
    document_count: int
    total_chunks: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class KnowledgeBaseListResponse(BaseModel):
    """知识库列表响应"""
    id: int
    name: str
    description: Optional[str]
    vector_store_type: VectorStoreType
    vector_store_config_id: Optional[int]
    embedding_provider: EmbeddingProvider
    embedding_model: EmbeddingModel
    document_count: int
    total_chunks: int
    is_public: bool
    created_at: datetime
    
    class Config:
        from_attributes = True


# ============= 文档 Schemas =============

class DocumentUpload(BaseModel):
    """文档上传（表单数据不需要 schema，这里用于元数据）"""
    pass


class DocumentResponse(BaseModel):
    """文档响应模型"""
    id: int
    knowledge_base_id: int
    filename: str
    file_type: str
    file_size: int
    status: DocumentStatus
    error_message: Optional[str]
    chunk_count: int
    character_count: int
    uploaded_at: datetime
    processed_at: Optional[datetime]
    
    class Config:
        from_attributes = True


class DocumentListResponse(BaseModel):
    """文档列表响应"""
    id: int
    filename: str
    file_type: str
    file_size: int
    status: DocumentStatus
    chunk_count: int
    character_count: int
    uploaded_at: datetime
    
    class Config:
        from_attributes = True


# ============= 文档块 Schemas =============

class ChunkResponse(BaseModel):
    """文档块响应"""
    id: int
    document_id: int
    content: str
    chunk_index: int
    metadata: Optional[str] = Field(None, alias='meta_data')
    created_at: datetime
    
    @property
    def character_count(self) -> int:
        """计算字符数"""
        return len(self.content) if self.content else 0
    
    class Config:
        from_attributes = True
        populate_by_name = True  # 允许使用 alias


# ============= 机器人关联知识库 Schemas =============

class RobotKnowledgeBaseAssociate(BaseModel):
    """关联知识库到机器人"""
    knowledge_base_ids: List[int] = Field(..., min_items=1)


class RobotKnowledgeBaseResponse(BaseModel):
    """机器人关联的知识库"""
    robot_id: int
    knowledge_bases: List[KnowledgeBaseListResponse]


# ============= 批量导入 Schemas =============

class BatchImportRequest(BaseModel):
    """批量导入请求"""
    directory_path: str = Field(..., description="本地目录路径")
    file_extensions: List[str] = Field(default=[".txt", ".pdf", ".docx"], description="要导入的文件扩展名")


class BatchImportResponse(BaseModel):
    """批量导入响应"""
    success_count: int
    failed_count: int
    total_count: int
    failed_files: List[str] = []


# ============= 检索相关 Schemas =============

class SearchRequest(BaseModel):
    """知识库搜索请求"""
    query: str = Field(..., min_length=1)
    top_k: int = Field(default=5, ge=1, le=20)
    knowledge_base_ids: Optional[List[int]] = None


class SearchResult(BaseModel):
    """搜索结果项"""
    content: str
    score: float
    document_id: int
    document_name: str
    chunk_index: int
    metadata: Optional[dict] = None


class SearchResponse(BaseModel):
    """搜索响应"""
    query: str
    results: List[SearchResult]
    total: int

